File: engine/next-record.js
#!/usr/bin/env node
'use strict';

/**
 * Issue directory picker: finds the next un-analyzed record from
 * deltas.ndjson and creates a prepared issue directory for it.
 *
 * Usage:
 *   node engine/next-record.js --job jobs/<round-name>
 */

const fs = require('fs');
const path = require('path');
const readline = require('readline');

function getArg(flag, def) {
  const i = process.argv.indexOf(flag);
  return i >= 0 && i + 1 < process.argv.length ? process.argv[i + 1] : def;
}

const JOB_DIR = getArg('--job', null);
if (!JOB_DIR) {
  console.error('Usage: node engine/next-record.js --job <job-directory>');
  process.exit(1);
}

const jobDir = path.resolve(JOB_DIR);
const { tolerances } = require(path.join(jobDir, 'tolerances'));
const DELTAS_FILE = path.join(jobDir, 'results/deltas/deltas.ndjson');
const ISSUES_DIR = path.join(jobDir, 'issues');

function sortKeysDeep(obj) {
  if (obj === null || obj === undefined || typeof obj !== 'object') return obj;
  if (Array.isArray(obj)) return obj.map(sortKeysDeep);
  const sorted = {};
  for (const key of Object.keys(obj).sort()) {
    sorted[key] = sortKeysDeep(obj[key]);
  }
  return sorted;
}

function runTolerancePipeline(record) {
  let prod, dev;
  try { prod = JSON.parse(record.prodBody); } catch { prod = null; }
  try { dev = JSON.parse(record.devBody); } catch { dev = null; }

  const ctx = { record, prod, dev };
  const applied = [];
  for (const t of tolerances) {
    const action = t.match(ctx);
    if (action === 'skip') {
      applied.push(`${t.id}: skip`);
      return { prod: ctx.prod, dev: ctx.dev, skippedBy: t.id, applied };
    }
    if (action === 'normalize' && ctx.prod && ctx.dev) {
      const result = t.normalize(ctx);
      ctx.prod = result.prod;
      ctx.dev = result.dev;
      applied.push(`${t.id}: normalize`);
    }
  }
  return { prod: sortKeysDeep(ctx.prod), dev: sortKeysDeep(ctx.dev), skippedBy: null, applied };
}

async function main() {
  if (!fs.existsSync(DELTAS_FILE)) {
    console.error(`Delta file not found: ${DELTAS_FILE}`);
    process.exit(1);
  }

  fs.mkdirSync(ISSUES_DIR, { recursive: true });

  const rl = readline.createInterface({
    input: fs.createReadStream(DELTAS_FILE),
    crlfDelay: Infinity,
  });

  let total = 0;
  let analyzed = 0;
  let found = null;

  for await (const line of rl) {
    const raw = line.trim();
    if (!raw) continue;
    total++;

    const record = JSON.parse(raw);
    const recordId = record.id;
    if (!recordId) continue;

    const analysisFile = path.join(ISSUES_DIR, recordId, 'analysis.md');

    if (fs.existsSync(analysisFile)) {
      analyzed++;
      continue;
    }

    if (!found) {
      found = { raw, recordId, lineno: total };
    }
  }

  if (!found) {
    if (total === 0) {
      console.error(`Delta file is empty: ${DELTAS_FILE}`);
    } else {
      console.error(`All ${total} records have been analyzed!`);
    }
    process.exit(1);
  }

  // Create the issue directory and write files
  const { raw, recordId, lineno } = found;
  const record = JSON.parse(raw);
  const issueDir = path.join(ISSUES_DIR, recordId);
  fs.mkdirSync(issueDir, { recursive: true });

  // Write record.json
  fs.writeFileSync(
    path.join(issueDir, 'record.json'),
    JSON.stringify(record, null, 2)
  );

  // Write prod-raw.json and dev-raw.json
  let prodRaw, devRaw;
  try { prodRaw = JSON.parse(record.prodBody); } catch { prodRaw = null; }
  try { devRaw = JSON.parse(record.devBody); } catch { devRaw = null; }

  fs.writeFileSync(
    path.join(issueDir, 'prod-raw.json'),
    JSON.stringify(prodRaw, null, 2)
  );
  fs.writeFileSync(
    path.join(issueDir, 'dev-raw.json'),
    JSON.stringify(devRaw, null, 2)
  );

  // Run tolerance pipeline for normalized output
  const normalized = runTolerancePipeline(record);

  fs.writeFileSync(
    path.join(issueDir, 'prod-normalized.json'),
    JSON.stringify(normalized.prod, null, 2)
  );
  fs.writeFileSync(
    path.join(issueDir, 'dev-normalized.json'),
    JSON.stringify(normalized.dev, null, 2)
  );

  // Write applied tolerances
  fs.writeFileSync(
    path.join(issueDir, 'applied-tolerances.txt'),
    normalized.applied.length > 0
      ? normalized.applied.join('\n') + '\n'
      : '(none)\n'
  );

  // Write pick context so we can track delta count trajectory across rounds
  const remaining = total - analyzed;
  const priority = record.comparison?.priority || '?';
  fs.writeFileSync(
    path.join(issueDir, 'pick-context.json'),
    JSON.stringify({ pickedAt: new Date().toISOString(), total, analyzed, remaining, priority }, null, 2)
  );

  // Print summary
  console.log(`Record: ${lineno}/${total} (${analyzed} analyzed, ${remaining} remaining)`);
  console.log(`Priority: ${priority}`);
  console.log(`Issue dir: ${issueDir}`);
  console.log(`Record ID: ${record.id || '?'}`);
  console.log(`URL: ${record.url || '?'}`);
  console.log(`Method: ${record.method || '?'}`);
  console.log(`Prod status: ${record.prodStatus || '?'}`);
  console.log(`Dev status: ${record.devStatus || '?'}`);
  console.log(`Operation: ${record.comparison?.op || '?'}`);
  console.log(`Lookup: grep -n '${record.id}' ${path.join(jobDir, 'comparison.ndjson')}`);
}

main().catch(e => { console.error(e); process.exit(1); });

File: next-record.py
#!/usr/bin/env python3
"""Pick the next un-analyzed record from a priority delta file.

Reads the delta file sequentially and returns the first record whose
MD5 hash is not already in the analyzed file. Exits 1 if all records
have been analyzed.

Usage:
    python3 next-record.py --priority P6
    python3 next-record.py --input results/deltas/p6.ndjson --analyzed results/p6-analyzed.txt
"""
import sys
import argparse
import hashlib
import json
import os

SCRIPT_DIR = os.path.dirname(os.path.abspath(__file__))


def main():
    parser = argparse.ArgumentParser(description="Pick next un-analyzed triage record")
    parser.add_argument("--priority", help="Priority level (e.g. P6). Derives file paths automatically.")
    parser.add_argument("--input", help="Path to delta NDJSON file (overrides --priority)")
    parser.add_argument("--analyzed", help="Path to analyzed ledger file (overrides --priority)")
    args = parser.parse_args()

    if args.input and args.analyzed:
        delta_file = args.input
        analyzed_file = args.analyzed
    elif args.priority:
        tag = args.priority.lower()
        delta_file = os.path.join(SCRIPT_DIR, f"results/deltas/{tag}.ndjson")
        analyzed_file = os.path.join(SCRIPT_DIR, f"results/{tag}-analyzed.txt")
    else:
        parser.error("Provide either --priority or both --input and --analyzed")

    if not os.path.exists(delta_file):
        print(f"Delta file not found: {delta_file}", file=sys.stderr)
        sys.exit(1)

    # Load analyzed hashes
    analyzed = set()
    if os.path.exists(analyzed_file):
        with open(analyzed_file) as f:
            for line in f:
                line = line.strip()
                if line:
                    analyzed.add(line.split(":")[0].strip())

    # Read delta file sequentially, find first un-analyzed
    total = 0
    with open(delta_file) as f:
        for lineno_0, line in enumerate(f):
            raw = line.strip()
            if not raw:
                continue
            total += 1
            md5 = hashlib.md5(raw.encode()).hexdigest()
            if md5 not in analyzed:
                rec = json.loads(raw)
                record_id = rec.get("id", "?")
                # Count remaining (including this one)
                remaining = total - len(analyzed)
                # Count rest of file
                for rest_line in f:
                    if rest_line.strip():
                        total += 1
                remaining = total - len(analyzed)

                print(f"Record: {lineno_0 + 1}/{total} ({len(analyzed)} analyzed, {remaining} remaining)")
                print(f"MD5: {md5}")
                print(f"Record ID: {record_id}")
                print(f"URL: {rec.get('url', '?')}")
                print(f"Method: {rec.get('method', '?')}")
                print(f"Prod status: {rec.get('prodStatus', rec.get('prod', {}).get('status', '?'))}")
                print(f"Dev status: {rec.get('devStatus', rec.get('dev', {}).get('status', '?'))}")
                print(f"Lookup: grep -n '{record_id}' comparison.ndjson")
                print(f"---RECORD---")
                print(raw)
                sys.exit(0)

    if total == 0:
        print(f"Delta file is empty: {delta_file}", file=sys.stderr)
    else:
        print(f"All {total} records have been analyzed!", file=sys.stderr)
    sys.exit(1)


if __name__ == "__main__":
    main()
